# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/WebGL2ContextState.cpp
# Commit: ccfa767dba64
# Full Hash: ccfa767dba644dc193e0246eb8e8ff3377e8b8a5
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-01-09 03:40:18
# Regressor Bug: 1477756
# File Overlap Count: 1
# Description:
#   Bug 1477756 - Client-side bindings mirror for precise CC, and merge similar codepaths. r=handyman
#   
#   * Context loss using RAII
#   * Move Program reflection Client-side
#   
# ==============================================================================

diff -r 71c122ac0ca7 -r ccfa767dba64 dom/canvas/WebGL2ContextState.cpp
--- a/dom/canvas/WebGL2ContextState.cpp	Wed Jan 08 22:19:14 2020 +0000
+++ b/dom/canvas/WebGL2ContextState.cpp	Wed Jan 08 22:19:16 2020 +0000
@@ -16,7 +16,7 @@
 
 namespace mozilla {
 
-MaybeWebGLVariant WebGL2Context::GetParameter(GLenum pname) {
+Maybe<double> WebGL2Context::GetParameter(GLenum pname) {
   const FuncScope funcScope(*this, "getParameter");
   // The following cases are handled in WebGLContext::GetParameter():
   //     case LOCAL_GL_MAX_COLOR_ATTACHMENTS:
@@ -32,24 +32,21 @@
     case LOCAL_GL_SAMPLE_COVERAGE: {
       realGLboolean b = 0;
       gl->fGetBooleanv(pname, &b);
-      return AsSomeVariant(bool(b));
+      return Some(bool(b));
     }
 
     case LOCAL_GL_TRANSFORM_FEEDBACK_ACTIVE:
-      return AsSomeVariant(mBoundTransformFeedback->mIsActive);
+      return Some(mBoundTransformFeedback->mIsActive);
     case LOCAL_GL_TRANSFORM_FEEDBACK_PAUSED:
-      return AsSomeVariant(mBoundTransformFeedback->mIsPaused);
+      return Some(mBoundTransformFeedback->mIsPaused);
 
     /* GLenum */
     case LOCAL_GL_READ_BUFFER: {
-      if (!mBoundReadFramebuffer)
-        return AsSomeVariant(std::move(mDefaultFB_ReadBuffer));
+      if (!mBoundReadFramebuffer) return Some(mDefaultFB_ReadBuffer);
 
-      if (!mBoundReadFramebuffer->ColorReadBuffer())
-        return AsSomeVariant(LOCAL_GL_NONE);
+      if (!mBoundReadFramebuffer->ColorReadBuffer()) return Some(LOCAL_GL_NONE);
 
-      return AsSomeVariant(std::move(
-          mBoundReadFramebuffer->ColorReadBuffer()->mAttachmentPoint));
+      return Some(mBoundReadFramebuffer->ColorReadBuffer()->mAttachmentPoint);
     }
 
     case LOCAL_GL_FRAGMENT_SHADER_DERIVATIVE_HINT:
@@ -66,9 +63,7 @@
     case LOCAL_GL_MAX_SAMPLES:
     case LOCAL_GL_MAX_TEXTURE_LOD_BIAS:
     case LOCAL_GL_MAX_TRANSFORM_FEEDBACK_INTERLEAVED_COMPONENTS:
-    case LOCAL_GL_MAX_TRANSFORM_FEEDBACK_SEPARATE_ATTRIBS:
     case LOCAL_GL_MAX_TRANSFORM_FEEDBACK_SEPARATE_COMPONENTS:
-    case LOCAL_GL_MAX_UNIFORM_BUFFER_BINDINGS:
     case LOCAL_GL_MAX_VERTEX_OUTPUT_COMPONENTS:
     case LOCAL_GL_MAX_VERTEX_UNIFORM_BLOCKS:
     case LOCAL_GL_MAX_VERTEX_UNIFORM_COMPONENTS:
@@ -76,46 +71,39 @@
     case LOCAL_GL_PACK_ROW_LENGTH:
     case LOCAL_GL_PACK_SKIP_PIXELS:
     case LOCAL_GL_PACK_SKIP_ROWS:
-    case LOCAL_GL_UNIFORM_BUFFER_OFFSET_ALIGNMENT:
     case LOCAL_GL_UNPACK_IMAGE_HEIGHT:
     case LOCAL_GL_UNPACK_ROW_LENGTH: {
       GLint val;
       gl->fGetIntegerv(pname, &val);
-      return AsSomeVariant(val);
+      return Some(val);
     }
 
     case LOCAL_GL_UNPACK_SKIP_IMAGES:
-      return AsSomeVariant(mPixelStore.mUnpackSkipImages);
+      return Some(mPixelStore.mUnpackSkipImages);
 
     case LOCAL_GL_UNPACK_SKIP_PIXELS:
-      return AsSomeVariant(mPixelStore.mUnpackSkipPixels);
+      return Some(mPixelStore.mUnpackSkipPixels);
 
     case LOCAL_GL_UNPACK_SKIP_ROWS:
-      return AsSomeVariant(mPixelStore.mUnpackSkipRows);
-
-    case LOCAL_GL_MAX_3D_TEXTURE_SIZE:
-      return AsSomeVariant(mGLMax3DTextureSize);
-
-    case LOCAL_GL_MAX_ARRAY_TEXTURE_LAYERS:
-      return AsSomeVariant(mGLMaxArrayTextureLayers);
+      return Some(mPixelStore.mUnpackSkipRows);
 
     case LOCAL_GL_MAX_VARYING_COMPONENTS: {
       // On OS X Core Profile this is buggy.  The spec says that the
       // value is 4 * GL_MAX_VARYING_VECTORS
       GLint val;
       gl->fGetIntegerv(LOCAL_GL_MAX_VARYING_VECTORS, &val);
-      return AsSomeVariant(4 * val);
+      return Some(4 * val);
     }
 
     /* GLint64 */
     case LOCAL_GL_MAX_CLIENT_WAIT_TIMEOUT_WEBGL:
-      return AsSomeVariant(kMaxClientWaitSyncTimeoutNS);
+      return Some(kMaxClientWaitSyncTimeoutNS);
 
     case LOCAL_GL_MAX_ELEMENT_INDEX:
       // GL_MAX_ELEMENT_INDEX becomes available in GL 4.3 or via ES3
       // compatibility
       if (!gl->IsSupported(gl::GLFeature::ES3_compatibility))
-        return AsSomeVariant(UINT32_MAX);
+        return Some(UINT32_MAX);
 
       /*** fall through to fGetInteger64v ***/
       [[fallthrough]];
@@ -125,69 +113,16 @@
     case LOCAL_GL_MAX_UNIFORM_BLOCK_SIZE: {
       GLint64 val;
       gl->fGetInteger64v(pname, &val);
-      return AsSomeVariant(static_cast<double>(val));
+      return Some(static_cast<double>(val));
     }
 
     /* GLuint64 */
     case LOCAL_GL_MAX_SERVER_WAIT_TIMEOUT: {
       GLuint64 val;
       gl->fGetInteger64v(pname, (GLint64*)&val);
-      return AsSomeVariant(static_cast<double>(val));
+      return Some(static_cast<double>(val));
     }
 
-    case LOCAL_GL_COPY_READ_BUFFER_BINDING:
-      return AsSomeVariant(std::move(mBoundCopyReadBuffer.get()));
-
-    case LOCAL_GL_COPY_WRITE_BUFFER_BINDING:
-      return AsSomeVariant(std::move(mBoundCopyWriteBuffer.get()));
-
-    case LOCAL_GL_PIXEL_PACK_BUFFER_BINDING:
-      return AsSomeVariant(std::move(mBoundPixelPackBuffer.get()));
-
-    case LOCAL_GL_PIXEL_UNPACK_BUFFER_BINDING:
-      return AsSomeVariant(std::move(mBoundPixelUnpackBuffer.get()));
-
-    case LOCAL_GL_TRANSFORM_FEEDBACK_BUFFER_BINDING:
-      return AsSomeVariant(std::move(mBoundTransformFeedbackBuffer.get()));
-
-    case LOCAL_GL_UNIFORM_BUFFER_BINDING:
-      return AsSomeVariant(std::move(mBoundUniformBuffer.get()));
-
-    // DRAW_FRAMEBUFFER_BINDING is the same as FRAMEBUFFER_BINDING.
-    case LOCAL_GL_READ_FRAMEBUFFER_BINDING:
-      return AsSomeVariant(std::move(mBoundReadFramebuffer.get()));
-
-    case LOCAL_GL_SAMPLER_BINDING:
-      return AsSomeVariant(std::move(mBoundSamplers[mActiveTexture].get()));
-
-    case LOCAL_GL_TEXTURE_BINDING_2D_ARRAY:
-      return AsSomeVariant(
-          std::move(mBound2DArrayTextures[mActiveTexture].get()));
-
-    case LOCAL_GL_TEXTURE_BINDING_3D:
-      return AsSomeVariant(std::move(mBound3DTextures[mActiveTexture].get()));
-
-    case LOCAL_GL_TRANSFORM_FEEDBACK_BINDING: {
-      WebGLTransformFeedback* tf = mBoundTransformFeedback;
-      if (tf == mDefaultTransformFeedback) {
-        tf = nullptr;
-      }
-      return AsSomeVariant(std::move(tf));
-    }
-
-    case LOCAL_GL_VERTEX_ARRAY_BINDING: {
-      WebGLVertexArray* vao = (mBoundVertexArray != mDefaultVertexArray)
-                                  ? mBoundVertexArray.get()
-                                  : nullptr;
-      return AsSomeVariant(std::move(vao));
-    }
-
-    case LOCAL_GL_VERSION:
-      return AsSomeVariant(nsCString("WebGL 2.0"));
-
-    case LOCAL_GL_SHADING_LANGUAGE_VERSION:
-      return AsSomeVariant(nsCString("WebGL GLSL ES 3.00"));
-
     default:
       return WebGLContext::GetParameter(pname);
   }